keras_nlp/samplers/sampler.py

```diff
@@ -20,405 +20,129 @@
 from keras_nlp.api_export import keras_nlp_export
 from keras_nlp.utils.python_utils import format_docstring
 
-base_sampler_args_docstring = """
-    jit_compile: bool, defaults to True. If True, XLA compilation will be used.
-    run_eagerly: bool, defaults to False. If True, the sampler will run in
-        the eager mode.
-    """
-
 call_args_docstring = """
-    prompt: a list of integers or an integer Tensor, can be 1D or 2D. The
-        initial tokens to append generated tokens.
-    token_probability_fn: a function that generates the probability of
-        the next token over the whole vocabulary for each input token.
-    max_length: int. The max length of generated sequence.
-    mask: a tensor, defaults to None. The padding mask of the prompt.
-    end_token_id: int, defaults to None. The token marking the end of the
-        sequence, once encountered the generation is finished for the exact
-        sequence. If None, every sequence is generated up to `max_length`.
-        If set, all tokens after encountering `end_token_id` will be
-        replaced with `pad_token_id`.
-    from_logits: bool, defaults to True. Indicate if the `token_probability_fn`
-        returns logits. If False, `token_probability_fn` returns probability
-        distributions.
-    cache: a dense int tensor, a cache of intermediate key and value tensor
-        computed by each decoder self-attention layer. These values will only
-        be computed once for each new token in the generated sequence.
+    next: A function which takes in the `prompt, state, index` of the
+        current generation loop, and outputs a tuple `logits, state` with the
+        probability for the next token and state for the next iteration.
+    prompt: A 2D integer tensor with shape `(batch_size, max_length)`. This
+        tensor will be iteratively updated column by column with new sampled
+        values.
+    state: Optional. A tensor or nested structure of tensors that will be
```

Comment:
Okay... I finally understand how this state works in general. Correct me if I am wrong, sharing my understanding - this `state` is a free variable, how to use it is totally decided by users of `samplers`. Most often this is to hold cache, but it is in fact a backdoor people can utilize in the `next` function or when they want to override `__call__` method.

My thought is "state" is too broad for users to learn, also I am not clear on how to let it hold >=2 things, e.g., in contrastive search we need both cache and previous logits. As we cannot assume the existence of cache, how do we retrieve the previous logits robustly?

Reply:
Contrastive search is not covered by this case. IIUC contrastive search does not need a cache per say, or previous logits. What contrastive search needs is the hidden representation of every token to compute a cosine similarity metric. So to make contrastive search work, we will need to update the signature of next to something like this:

```
def next(prompt, state, index):
   return logits, dense, state
```

I can take a closer look at contrastive search implementation to understand what is out there. I was thinking to cover it as a follow up, but definitely worth some thought.

Reply:
I also think we might want to pull apart the user journeys here.

Someone writing a sampler is writing a drop in replacement for `sampler="beam"` or `sampler="greedy"`. `state` is not an important backdoor in this case, it's just more tensor variables that should be treated as loop variables. But from the perspective of the sampler writer, the `model` and it's forward pass are a black box.

In contrast, from the perspective of a model writer, the sampler should be a black box, where the user complies with the `next` contract and doesn't worry about how the sampling actually happens. Here, `state` can be used for introduction of arbitrary extra updatable state needed solely to compute the probability distribution of the next token. This could be the cache for a transformer decoder, the state state of a recurrent network, etc.

Overall, I am most interested in keeping the sampler simple and useful for our own purposes, but worth chatting through all this!